src/libstd/io/mod.rs

```diff
@@ -1549,93 +1594,182 @@ impl<R: Read> Iterator for Bytes<R> {
+                        _ => return Some(Err(Utf8CharsError::InvalidUtf8))
+                    }
                 }
             }
         }
-        Some(match str::from_utf8(&buf[..width]).ok() {
-            Some(s) => Ok(s.chars().next().unwrap()),
-            None => Err(CharsError::NotUtf8),
-        })
+
+        // Ranges can be checked against https://tools.ietf.org/html/rfc3629#section-4
+        let code_point = match core_str::utf8_char_width(first) {
+            1 => return Some(Ok(first as char)),
+            2 => {
+                let second = continuation_byte!(0x80...0xBF);
+                ((first & 0b0001_1111) as u32) << 6 | second
+            }
+            3 => {
+                let second = match first {
+                    0xE0        => continuation_byte!(0xA0...0xBF),
+                    0xE1...0xEC => continuation_byte!(0x80...0xBF),
+                    0xED        => continuation_byte!(0x80...0x9F),
+                    0xEE...0xEF => continuation_byte!(0x80...0xBF),
+                    _ => unreachable!(),
+                };
+                let third = continuation_byte!(0x80...0xBF);
+                ((first & 0b0000_1111) as u32) << 12 | second << 6 | third
+            }
+            4 => {
+                let second = match first {
+                    0xF0        => continuation_byte!(0x90...0xBF),
+                    0xF0...0xF3 => continuation_byte!(0x80...0xBF),
+                    0xF4        => continuation_byte!(0x80...0x8F),
+                    _ => unreachable!(),
+                };
+                let third = continuation_byte!(0x80...0xBF);
+                let fourth = continuation_byte!(0x80...0xBF);
+                ((first & 0b0000_0111) as u32) << 18 | second << 12 | third << 6 | fourth
+            }
+            _ => return Some(Err(Utf8CharsError::InvalidUtf8))
+        };
+        unsafe {
+            Some(Ok(char::from_u32_unchecked(code_point)))
+        }
     }
 }
 
-#[unstable(feature = "io", reason = "awaiting stability of Read::chars",
+#[unstable(feature = "io", reason = "awaiting stability of BufRead::utf8_chars",
            issue = "27802")]
-impl std_error::Error for CharsError {
+impl std_error::Error for Utf8CharsError {
     fn description(&self) -> &str {
         match *self {
-            CharsError::NotUtf8 => "invalid utf8 encoding",
-            CharsError::Other(ref e) => std_error::Error::description(e),
+            Utf8CharsError::InvalidUtf8 => "invalid UTF-8 byte sequence",
+            Utf8CharsError::IncompleteUtf8 => {
+                "stream ended in the middle of an UTF-8 byte sequence"
+            }
+            Utf8CharsError::Io(ref e) => std_error::Error::description(e),
         }
     }
     fn cause(&self) -> Option<&std_error::Error> {
         match *self {
-            CharsError::NotUtf8 => None,
-            CharsError::Other(ref e) => e.cause(),
+            Utf8CharsError::InvalidUtf8 | Utf8CharsError::IncompleteUtf8 => None,
```

Comment:
Why? I feel like there is no point in doing that. Two common strategies for dealing with character decoding errors are replacing them with U+FFFD (`utf8_chars_lossy` does this) and aborting the operation on the first error. Neither care about the bytes in an invalid sequence.

The only case where it might be useful is Python’s `surrogateescape` strategy, but that doesn’t work in Rust since `char` disallows surrogates code points. To support sometimes-UTF-8 arbitrary bytes in Rust, I’d recommend using `[u8]` / `Vec<u8>` and some methods like `to_str(&self) -> Option<&str>`. This is what the standard library does on Unix for `OsStr` / `OsString`.